src/modules/job-manager/update.c

```diff
@@ -163,7 +178,147 @@ static void update_job (struct job_manager *ctx,
         flux_log_error (ctx->h, "%s: flux_respond", __FUNCTION__);
     return;
 error:
-    if (flux_respond_error (ctx->h, msg, errno, error ? error : errstr) < 0)
+    if (flux_respond_error (ctx->h, msg, errno, errstr) < 0)
+        flux_log_error (ctx->h, "%s: flux_respond_error", __FUNCTION__);
+}
+
+static void feasibility_cb (flux_future_t *f, void *arg)
+{
+    struct update_request *req = arg;
+    if (flux_future_get (f, NULL) < 0) {
+        if (flux_respond_error (req->update->ctx->h,
+                                req->msg,
+                                errno,
+                                future_strerror (f, errno)) < 0)
+            flux_log_error (req->update->ctx->h,
+                            "%s: flux_respond_error",
+                            __FUNCTION__);
+    }
+    else {
+        post_job_updates (req->update->ctx,
+                          req->msg,
+                          req->cred,
+                          req->job,
+                          req->updates,
+                          req->validate);
+    }
+    zlistx_delete (req->update->pending_requests, req->handle);
+    flux_future_destroy (f);
+}
+
+static struct update_request *
+pending_request_create (struct update *update,
+                        const flux_msg_t *msg,
+                        struct flux_msg_cred cred,
+                        struct job *job,
+                        json_t *updates,
+                        int validate)
+{
+    struct update_request *req = NULL;
+    if (!(req = update_request_create (update, msg, cred, job, updates))
+        || !(req->handle = zlistx_add_end (update->pending_requests, req)))
+        goto error;
+    req->validate = validate;
+    return req;
+error:
+    update_request_destroy (req);
+    errno = ENOMEM;
+    return NULL;
+}
+
+static int update_feasibility_check (struct update *update,
+                                     const flux_msg_t *msg,
+                                     struct flux_msg_cred cred,
+                                     struct job *job,
+                                     json_t *updates,
+                                     int validate)
+{
+    json_t *jobspec = NULL;
+    struct update_request *req = NULL;
+    flux_future_t *f = NULL;
+
+    if (!(jobspec = job_jobspec_with_updates (job, updates))
+        || !(req = pending_request_create (update,
+                                           msg,
+                                           cred,
+                                           job,
+                                           updates,
+                                           validate))
+        || !(f = flux_rpc_pack (update->ctx->h,
+                                "sched.feasibility",
+                                0,
+                                0,
+                                "{s:O}",
+                                "jobspec", jobspec))
+        || flux_future_then (f, -1., feasibility_cb, req) < 0)
+        goto error;
```

Comment:
Would the future be leaked if job-manager were unloaded with a feasibility check in progress?

Reply:
Ah, yeah. The whole reason for the pending requests list was to do this cleanup, then I forgot to add the future there! :facepalm: